### PR TITLE
all: run Docker dependent tests whenever possible

### LINKED ISSUE #2005: testing: explore options to check local container status before running tests
We discussed in the vault [PR](https://github.com/google/go-cloud/pull/1983#discussion_r281254551) that we want to get a sync point after localXXX.sh boots a container of a certain server for testing.

There are a few options:
1. Do nothing more, we set a timeout on the constructors and fail if the server is not up after a few seconds.
1. Wait a few seconds after the container starts and check its health.
1. If the previous one is still flaky, write a Go program to do more proper waiting.

### PR DESCRIPTION
**Pull request by shantuo**
Fixes #2005 

All tests that require Docker container to run will run when either:
 1) Not on Travis.
 2) On Travis Linux environment, where Docker is available.

### CODE REVIEW

File: secrets/vault/vault_test.go [hunk 1]
```diff
@@ -37,6 +39,8 @@ const (
 	testToken  = "faketoken"
 )
 
+var enableTransit sync.Once
```
Comment (eliben) [inline secrets/vault/vault_test.go]: Document this var

File: internal/testing/setup/setup.go [hunk 2]
```diff
@@ -334,3 +334,11 @@ func NewGCPDialOptions(t *testing.T, recording bool, filename string) (opts []gr
 	}
 	return opts, done
 }
+
+// RunTestsDependingOnDocker returns true when either:
+// 1) Not on Travis.
+// 2) On Travis Linux environment, where Docker is available.
+func RunTestsDependingOnDocker() bool {
```
Comment (eliben) [inline internal/testing/setup/setup.go]: I feel the name of this function could be better, conveying it's a predicate. Something like `HasDockerTestEnvironment` or `CanRunLocalServerTests`, etc.

**Comment by eliben** [comment]
OOC, how do you verify these tests actually ran on Travis, rather than being skipped?

File: secrets/vault/vault_test.go (hunk 1 above)
Comment (shantuo) [reply secrets/vault/vault_test.go]: Done.

File: internal/testing/setup/setup.go (hunk 2 above)
Comment (shantuo) [reply internal/testing/setup/setup.go]: Went with `HasDockerTestEnvironment`

File: pubsub/kafkapubsub/kafka_test.go [hunk 3]
```diff
@@ -53,21 +54,21 @@ type harness struct {
 }
 
 var checkOnce sync.Once
-var kafkaRunning bool
+var kafkaError error
```
Comment (vangent) [inline pubsub/kafkapubsub/kafka_test.go]: I think the `sync.Once` and `kafkaError` variables and `localKafkaRunning` can all be dropped. The tests will just fail naturally if Kafka isn't running.

File: pubsub/kafkapubsub/kafka_test.go [hunk 4]
```diff
@@ -196,8 +197,11 @@ func (asTest) BeforeSend(as func(interface{}) bool) error {
 
 // TestKafkaKey tests sending/receiving a message with the Kafka message key set.
 func TestKafkaKey(t *testing.T) {
-	if !localKafkaRunning() {
-		t.Skip("No local Kafka running, see pubsub/kafkapubsub/localkafka.sh")
+	if !setup.HasDockerTestEnvironment() {
+		t.Skip("Skipping Kafka tests since the Kafka server is not available")
+	}
+	if err := localKafkaRunning(); err != nil {
```
Comment (vangent) [inline pubsub/kafkapubsub/kafka_test.go]: Same here, not needed.

File: runtimevar/etcdvar/etcdvar_test.go [hunk 5]
```diff
@@ -38,20 +39,25 @@ var (
...
 func init() {
-	etcdClient, etcdError = clientv3.NewFromURL("http://localhost:2379")
-	if etcdError == nil {
-		// Check to see if the local etcd is actually running.
-		ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
-		defer cancel()
-		_, etcdError = etcdClient.Put(ctx, "unused", "unused")
+	if setup.HasDockerTestEnvironment() {
+		etcdClient, etcdError = clientv3.NewFromURL("http://localhost:2379")
+		if etcdError == nil {
```
Comment (vangent) [inline runtimevar/etcdvar/etcdvar_test.go]: Lines 44-48 can be dropped; same as above, tests will just fail naturally if etcd isn't running.

File: secrets/vault/vault_test.go [hunk 6]
```diff
@@ -37,6 +39,9 @@ const (
 	testToken  = "faketoken"
 )
 
+// enableTransit cheks and makes sure the Transit API is enable only once.
```
Comment (vangent) [inline secrets/vault/vault_test.go]: nit: cheks -> checks, enable -> enabled

**Comment by eliben** [comment]
> I ran locally with the travis env, our Travis script doesn't have `-v` for `go test` so can't tell from there.

I think it's worthwhile to test this at least once (like add `-v` in your PR and remove it later)?

File: secrets/vault/vault_test.go (hunk 6 above)
Comment (shantuo) [reply secrets/vault/vault_test.go]: Done.

**Pull request merged**